doc/designs/quic-design/synthesis.md

```diff
@@ -0,0 +1,494 @@
+ *
+ * After creating a SSL_CTX with one of the above methods, an SSL object is
+ * created:
+ *
+ *   ssl = SSL_new(ctx);
+ *
+ * This is considered a QUIC connection object from the outset,
+ * i.e. SSL_is_connection() returns 1 immediately. The connection
+ * has not yet been made.
+ *
+ * SSL_set_connect_state(ssl); must be called. It is an error to call
+ * SSL_set_accept_state(ssl); with one of the above methods.
+ *
+ * It is necessary that SSL_set_tlsext_host_name() is called as SNI is
+ * mandatory. SSL_set1_host() should be called to allow certificate
+ * verification. Later we might provide a single call to set both of these to be
+ * set ala SSL_connect_ex these but this can wait.
+ *
+ * At this point:
+ *
+ *   If SSL_set_fd is called the SSL object, which represents a QUIC
+ *   connection, we take it as a cue that it must use this FD.
+ *
+ *   If the FD has a destination (this can be tested via getpeername())
+ *   this can be used as-is. We might always use sendto though and
+ *   change the address we use later if connection migration occurs.
+ *
+ *   If the FD has no destination we need to determine a destination.
+ *   We can either fail in this case or do a DNS lookup ourselves.
+ *   We already have the infrastructure for this with BIO_lookup.
+ *   In the BIO_new_ssl_connect case the lookup will already have been
+ *   done by BIO_s_connect.
+ *
+ *   If SSL_set_bio is called the SSL object, which represents a QUIC
+ *   connection, takes it as a cue that it must use this BIO, and that the BIO
+ *   has datagram semantics. If the BIO supports sendto/recvfrom/etc. we might
+ *   be able to support connection migration, otherwise it is e.g. a
+ *   BIO_dgram_mem (e.g. DDD demo 05) which has not indicated that sendto
+ *   addresses will be honoured and migration is disabled.
+ *
+ * SSL_connect(ssl) may be called or it may be implied by trying to
+ * use SSL_read/SSL_write.
+ *
+ *
+ * Connection Establishment: BIO_new_ssl_connect
+ * =============================================
+ *
+ * BIO_new_ssl_connect will be changed to create a SOCK_DGRAM stream
+ * when the SSL_CTX passed to it uses a QUIC method.
+ *
+ * At some point in the future (beyond the MVP) we can change
+ * BIO_new_ssl_connect to use alternate methods to provide target addresses to
+ * the QUIC connection object it creates, to the extent necessary to support
+ * connection migration, etc. This abstracts these changes from codebases which
+ * use BIO_new_ssl_connect.
+ */
+
+/*
+ * Get the polling FD for the BIO. This might be different from the value
+ * returned by BIO_get_fd and should be used by the application for polling. For
+ * BIO_f_ssl it uses SSL_get_poll_fd; for anything else it will just forward to
+ * BIO_get_fd.
+ */
+__owur int BIO_get_poll_fd(BIO *bio, int *fd); /* BIO_C_GET_POLL_FD */
+__owur int SSL_get_poll_fd(SSL *ssl, int *fd);
+
+/*
+ * Returns the number of milliseconds from now to the earliest (possible) timer
```

Comment:
Is milliseconds the right granularity? And if not, we might need a 64 bit return value type.

Reply:
I'd suggest _μs_ to align with `struct timeval`.  On Gbit networks, a _ms_ is a loooong time.  I would entertain argument for/against _ns_ too.

Reply:
A 64-bit time value gives us quite some time regardless...

With *ns*, it wraps after ~584 years (unsigned)
With *µs*, it wraps after ~600000 years (unsigned)

What to choose, what to choose...

Reply:
_ps_ would give us more room for future performance improvements.
[Plank time](https://en.wikipedia.org/wiki/Planck_units#Planck_time) would be completely future proof but 64 bits wouldn't be adequate.

Reply:
I think ns precision would be purely fictional. So I'd be for μs.

Reply:
I think you should go with ns, to match functions like nanosleep() that take a timespec containing ns. It should probably just return a timespec.

Reply:
This is still unresolved